src/python/gem5/resources/resource.py

```diff
@@ -753,13 +770,43 @@ def __init__(
             resource_version=resource_version,
         )
 
+    def _get_all_workload_json(self) -> List[Dict[str, str]]:
+        """
+        This function gets all the resources for the workload.
+
+        :returns: A list of all the resources for the workload.
+        """
+
+        mongo_query = [
+            {"id": resource_info[0], "resource_version": resource_info[1]}
+            for resource_info in self._workloads.keys()
+        ]
+        return get_multiple_resource_json_obj(
+            mongo_query, self._clients, self._gem5_version
+        )
+
     def __iter__(self) -> Generator["WorkloadResource", None, None]:
         """
         Returns a generator that iterates over the workloads in the suite.
 
         :yields: A generator that iterates over the workloads in the suite.
         """
-        yield from self._workloads.keys()
+
+        # We get all the workload objects from the database
+        workload_objects = self._get_all_workload_json()
+
+        for workload in workload_objects:
+            # We create a WorkloadResource object for each workload
+            # only when it is requested.
+            yield WorkloadResource(
+                local_path=self._local_path,
+                resource_directory=self._resource_directory,
+                download_md5_mismatch=self._download_md5_mismatch,
+                clients=self._clients,
+                gem5_version=self._gem5_version,
+                quiet=self._quiet,
+                **workload,
```

Comment:
I don't understand this line of code. I guess I don't know what's in the `workload` dictionary.

This is clever, but maybe not readable. A comment could help, or maybe the code could be be more verbose.